### PR TITLE
 Create dotc -decompiler

### PR DESCRIPTION
**Pull request by nicolasstucki**
Create infrastructure to support `dotp` tasty decompiler. The implementation of the decompiler is a tasty compiler that compiles to a text representation of the loaded code (currently with Tree.show).

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Needs rebase on #3436

File: compiler/src/dotty/tools/dotc/config/ScalaSettings.scala [hunk 1]
```diff
@@ -96,6 +96,7 @@ class ScalaSettings extends Settings.SettingGroup {
   val YnoPatmatOpt = BooleanSetting("-Yno-patmat-opt", "disable all pattern matching optimizations.")
   val YplainPrinter = BooleanSetting("-Yplain-printer", "Pretty-print using a plain printer.")
   val YprintSyms = BooleanSetting("-Yprint-syms", "when printing trees print info in symbols instead of corresponding info in trees.")
+  val YprintNoPrivate = BooleanSetting("-Y-print-no-private", "when printing trees print do not print private/protected members.")
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/config/ScalaSettings.scala]: should -Yprint, not -Y-print

**Comment by nicolasstucki** [comment]
Rebased and squashed

File: compiler/src/dotty/tools/dotc/decompiler/DecompilationPrinter.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    val doubleLine = "=" * pageWidth
+    val line = "-" * pageWidth
+
+    println(doubleLine)
+    println(unit.source)
+    println(line)
+
+    val code = unit.tpdTree.show
+    println(if (ctx.useColors) printing.SyntaxHighlighting(code) else code)
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/decompiler/DecompilationPrinter.scala]: It should really be `printing.SyntaxHighlighting(code).mkString`. We shouldn't assume that the underlying `Iterable[Char].toString` works as we expect.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/decompiler/DecompilationPrinter.scala]: I will remove the syntax highlighting as it is not compatible with all the output formats of `show`. I will work on coloring in a separate PR.

**Pull request merged**